File: Code/Components/DanpheEMR.ServerModel/BillingModels/Config/BillServiceItemModel.cs
using DanpheEMR.ServerModel.BillingModels;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace DanpheEMR.ServerModel
{
    public class BillServiceItemModel
    {
        [Key]
        public int ServiceItemId { get; set; }
        public int ServiceDepartmentId { get; set; }
        public int IntegrationItemId { get; set; }
        public string IntegrationName { get; set; }
        public string ItemCode { get; set; } //yub 24th sept '18
        public string ItemName { get; set; }
        public bool IsTaxApplicable { get; set; }
        public string Description { get; set; }
        public int? DisplaySeq { get; set; }//sud: 26July'18
        public bool IsDoctorMandatory { get; set; } //yub 24th sept '18
        public bool IsOT { get; set; }
        public bool IsProc { get; set; }
        public int? ServiceCategoryId { get; set; } //new column (need to have a FK relation with ServiceCategory table)
        public bool AllowMultipleQty { get; set; } //pratik 18th oct '19
        public string DefaultDoctorList { get; set; }
        public bool IsValidForReporting { get; set; }//pratik:7Aug2020
        public bool IsErLabApplicable { get; set; }//pratik:9Feb'21
        public int CreatedBy { get; set; }
        public DateTime CreatedOn { get; set; }
        public DateTime? ModifiedOn { get; set; }
        public int? ModifiedBy { get; set; }
        public bool IsActive { get; set; }

        public bool IsIncentiveApplicable { get; set; }

        //Krishna, 16thFeb'23 We might need these NotMapped properties, Need to check their references and proceed accordingly
        [NotMapped]
        public string ServiceDepartmentName { get; set; }
        [NotMapped]
        public List<BillMapPriceCategoryServiceItemModel> BilCfgItemsVsPriceCategoryMap { get; set; }



    }

    public class BillItemPrice_Old_16thFeb_23
    {
        public int BillItemPriceId { get; set; }
        public int ServiceDepartmentId { get; set; }
        public int ItemId { get; set; }
        public string ItemName { get; set; }
        public string ProcedureCode { get; set; }
        public double? Price { get; set; }
        public int? CreatedBy { get; set; }
        public DateTime? CreatedOn { get; set; }
        public DateTime? ModifiedOn { get; set; }
        public int? ModifiedBy { get; set; }
        public bool? IsActive { get; set; }
        public string IntegrationName { get; set; }

        public bool? TaxApplicable { get; set; }
        public string Description { get; set; }
        public bool? DiscountApplicable { get; set; }

        public bool? IsDoctorMandatory { get; set; } //yub 24th sept '18
        public string ItemCode { get; set; } //yub 24th sept '18

        public int? DisplaySeq { get; set; }//sud: 26July'18

        public bool? HasAdditionalBillingItems { get; set; } //Hom: 23 Jan'19
        public bool InsuranceApplicable { get; set; }
        public double? GovtInsurancePrice { get; set; }
        public bool IsInsurancePackage { get; set; }

        public bool? IsFractionApplicable { get; set; } //mahesh 8th feb'19


        public double? EHSPrice { get; set; }//sud:24Feb'19--New prices required.
        public double? SAARCCitizenPrice { get; set; }//sud:24Feb'19--New prices required.
        public double? ForeignerPrice { get; set; }//sud:24Feb'19--New prices required.
        public double? InsForeignerPrice { get; set; }//Pratik:8Nov'19--New prices required.

        public bool? IsNormalPriceApplicable { get; set; }//sud:18Apr'19--For Normal price 
        public bool? IsEHSPriceApplicable { get; set; }//sud:18Apr'19--For Normal price 
        public bool? IsForeignerPriceApplicable { get; set; }//sud:18Apr'19--For Normal price 
        public bool? IsSAARCPriceApplicable { get; set; }//sud:18Apr'19--For Normal price 
        public bool? IsInsForeignerPriceApplicable { get; set; }//Pratik:8Nov'19--For Normal price 

        public bool? IsOT { get; set; }
        public bool? IsProc { get; set; }
        public string Category { get; set; }

        public bool? AllowMultipleQty { get; set; } //pratik 18th oct '19
        public string DefaultDoctorList { get; set; }
        public bool? IsValidForReporting { get; set; }//pratik:7Aug2020
        public bool? IsZeroPriceAllowed { get; set; }//pratik:28Jan'21
        public bool? IsErLabApplicable { get; set; }//pratik:9Feb'21


        [NotMapped]
        public string ServiceDepartmentName { get; set; }
        [NotMapped]
        public List<BillMapPriceCategoryServiceItemModel> BilCfgItemsVsPriceCategoryMap { get; set; }
        public bool? IsPriceChangeAllowed { get; set; } //Krishna: 2ndNov'22

        //Sud:13Jan'22--initializing the constructor with Default Values.
        public BillItemPrice_Old_16thFeb_23()
        {
            IsPriceChangeAllowed = true;
        }

    }

}

File: Code/Websites/DanpheEMR/wwwroot/DanpheApp/src/app/settings-new/billing/shared/bill-service-item.model.ts
import {
  FormBuilder, FormControl, FormGroup, Validators
} from '@angular/forms';

export class BillServiceItemModel {
  public ServiceDepartmentId: number = null;
  public ItemName: string = null;
  public ServiceItemId: number = 0;
  public ItemCode: string = null;
  public Price: number = 0;
  public IntegrationItemId: number = null;
  public IntegrationName: string = null;
  public IsTaxApplicable: boolean = false;
  public Description: string = null;
  public DisplaySeq: number = 100;
  public IsDoctorMandatory: boolean = false;
  public IsOT: boolean = false;
  public IsProc: boolean = false;
  public ServiceCategoryId: number;
  public AllowMultipleQty: boolean = true;
  public DefaultDoctorList: string = null;
  public IsValidForReporting: boolean = false;
  public ModifiedBy: number = null;
  public IsErLabApplicable: boolean = false;
  public CreatedBy: number = null;
  public CreatedOn: string = null;
  public ModifiedOn: string = null;
  public IsActive: boolean = true;
  public ItemId: number = 0;
  public BillingItemValidator: FormGroup = null;
  public ServiceDepartmentName: string = null;
  public ServiceCategoryName: string = null;
  public DiscountApplicable: boolean = true;
  public BilCfgItemsVsPriceCategoryMap: Array<BillServiceItemsPriceCategoryMap> = new Array<BillServiceItemsPriceCategoryMap>();

  constructor() {

    const _formBuilder = new FormBuilder();
    this.BillingItemValidator = _formBuilder.group({
      'ServiceDepartmentId': ['', Validators.compose([Validators.required])],
      'ServiceCategoryId': ['', Validators.compose([Validators.required])],
      'ItemCode': ['', Validators.compose([Validators.maxLength(10)])],
      'ItemName': ['', Validators.compose([Validators.required, Validators.maxLength(100)])],
      'IntegrationName': ['', Validators.compose([Validators.required])],
    });
  }

  public IsDirty(fieldName): boolean {
    if (fieldName == undefined)
      return this.BillingItemValidator.dirty;
    else
      return this.BillingItemValidator.controls[fieldName].dirty;
  }

  public IsValid(): boolean { if (this.BillingItemValidator.valid) { return true; } else { return false; } }
  public IsValidCheck(fieldName, validator): boolean {
    if (fieldName == undefined) {
      return this.BillingItemValidator.valid;
    }

    else
      return !(this.BillingItemValidator.hasError(validator, fieldName));
  }


  public UpdateValidator(onOff: string, formControlName: string) {
    if (formControlName == "GovtInsurancePrice" && onOff == "on") {
      this.BillingItemValidator.controls['GovtInsurancePrice'].validator = Validators.compose([Validators.required]);
    }
    else {
      this.BillingItemValidator.controls['GovtInsurancePrice'].validator = Validators.compose([]);
    }
    this.BillingItemValidator.controls[formControlName].updateValueAndValidity();
  }


  public EnableControl(formControlName: string, enabled: boolean) {
    let currCtrol = this.BillingItemValidator.controls[formControlName];
    if (currCtrol) {
      if (enabled) {
        currCtrol.enable();
      }
      else {
        currCtrol.disable();
      }
    }
  }

  positiveNumberValdiator(control: FormControl): { [key: string]: boolean } {
    if (control) {
      if (control.value < 0)
        return { 'invalidNumber': true };
    }
  }

}
export class BillServiceItemsPriceCategoryMap {
  public PriceCategoryServiceItemMapId: number = 0;
  public PriceCategoryId: number = null;
  public Price: number = 0;
  public IsDiscountApplicable: boolean = false;
  public ItemLegalCode: string = "";
  public ItemLegalName: string = "";
  public IsSelected: boolean = false;
  public PriceCategoryName: string = "";
  public IsZeroPriceAllowed: boolean = false;
  public IsIncentiveApplicable: boolean = false;
  public HasAdditionalBillingItems: boolean = false;
  public IsPriceChangeAllowed: boolean = false;
  public ServiceItemId: number = 0;
  public ServiceDepartmentId: number = 0;
  public IsActive: boolean = false;
}

export class ServiceCategories {
  public ServiceCategoryId: number = 0;
  public ServiceCategoryName: string = null;
}


